manual/src/tutorials/parallelism.etex

```diff
@@ -0,0 +1,732 @@
+
+module T = Domainslib.Task
+
+let eval_A_times_u pool u v =
+  let n = Array.length v - 1 in
+  T.parallel_for pool ~start:0 ~finish:n ~body:(fun i ->
+    let vi = ref 0. in
+    for j = 0 to n do vi := !vi +. eval_A i j *. u.(j) done;
+    v.(i) <- !vi
+  )
+
+let eval_At_times_u pool u v =
+  let n = Array.length v - 1 in
+  T.parallel_for pool ~start:0 ~finish:n ~body:(fun i ->
+    let vi = ref 0. in
+    for j = 0 to n do vi := !vi +. eval_A j i *. u.(j) done;
+    v.(i) <- !vi
+  )
+
+let eval_AtA_times_u pool u v =
+  let w = Array.make (Array.length u) 0.0 in
+  eval_A_times_u pool u w; eval_At_times_u pool w v
+
+let () =
+  let pool = T.setup_pool ~num_additional_domains:(num_domains - 1) () in
+  let u = Array.make n 1.0  and  v = Array.make n 0.0 in
+  T.run pool (fun _ ->
+  for _i = 0 to 9 do
+    eval_AtA_times_u pool u v; eval_AtA_times_u pool v u
+  done);
+
+  let vv = ref 0.0  and  vBv = ref 0.0 in
+  for i=0 to n-1 do
+    vv := !vv +. v.(i) *. v.(i);
+    vBv := !vBv +. u.(i) *. v.(i)
+  done;
+  T.teardown_pool pool;
+  Printf.printf "%0.9f\n" (sqrt(!vBv /. !vv))
+\end{verbatim}
+
+Observe that the "parallel_for" function is isomorphic to the for-loop in the
+sequential version. No other change is required except for the boiler plate
+code to set up and tear down the pools.
+
+\begin{verbatim}
+$ ocamlopt -o spectralnorm.exe spectralnorm.ml
+$ ocamlfind ocamlopt -package domainslib -linkpkg -o spectralnorm_par.exe \
+  spectralnorm_par.ml
+$ hyperfine './spectralnorm.exe 4096' './spectralnorm_par.exe 2 4096' \
+            './spectralnorm_par.exe 4 4096' './spectralnorm_par.exe 8 4096'
+Benchmark 1: ./spectralnorm.exe 4096
+  Time (mean ± sd):     1.989 s ±  0.013 s    [User: 1.972 s, System: 0.007 s]
+  Range (min … max):    1.975 s …  2.018 s    10 runs
+
+Benchmark 2: ./spectralnorm_par.exe 2 4096
+  Time (mean ± sd):     1.083 s ±  0.015 s    [User: 2.140 s, System: 0.009 s]
+  Range (min … max):    1.064 s …  1.102 s    10 runs
+
+Benchmark 3: ./spectralnorm_par.exe 4 4096
+  Time (mean ± sd):    698.7 ms ±  10.3 ms    [User: 2730.8 ms, System: 18.3 ms]
+  Range (min … max):   680.9 ms … 721.7 ms    10 runs
+
+Benchmark 4: ./spectralnorm_par.exe 8 4096
+  Time (mean ± sd):    921.8 ms ±  52.1 ms    [User: 6711.6 ms, System: 51.0 ms]
+  Range (min … max):   838.6 ms … 989.2 ms    10 runs
+
+Summary
+  './spectralnorm_par.exe 4 4096' ran
+    1.32 ± 0.08 times faster than './spectralnorm_par.exe 8 4096'
+    1.55 ± 0.03 times faster than './spectralnorm_par.exe 2 4096'
+    2.85 ± 0.05 times faster than './spectralnorm.exe 4096'
+\end{verbatim}
+
+On the author's machine, the program scales well up to 4 domains but performs
+worse with 8 domains. This is due to the fact that unlike the parallel
+Fibonacci program, this program performs floating-point computations. At 8
+domains, the hyper-threaded siblings contend for the shared floating-point
+units on a core leading to worse behaviour than at 4 domains.
```

Comment:
I'm not convinced by this analysis, because even for 2 and 4 domains the speedup is not so good (1.83 and 2.16, respectively).  My explanation is rather that the tasks are smaller, so that the synchronisation is more expensive on the second example.

Reply:
>  My explanation is rather that the tasks are smaller,

`parallel_for` uses divide-and-conquer to distribute the work: https://github.com/ocaml-multicore/domainslib/pull/16. By default, each task should have enough chunked work to do. I would think that the tasks being small is not the issue. 

I explored this a bit more. I thought the issue was due to the impact of minor GCs. Compared to the parallel fibonacci example, this example allocates much more in the minor heap. 

```bash
kc@KCs-MacBook-Pro temp % OCAMLRUNPARAM="v=0x400" ./spectralnorm_par.exe 8 8192
1.274224153
allocated_words: 5369082705
minor_words: 5368902430
promoted_words: 96098
major_words: 276373
minor_collections: 25720
major_collections: 406
forced_major_collections: 0
heap_words: 249865
top_heap_words: 69641
mean_space_overhead: 867.869048
kc@KCs-MacBook-Pro temp % OCAMLRUNPARAM="v=0x400" ./fib_par2.exe 8 42
fib(42) = 433494437
allocated_words: 6864229
minor_words: 6864196
promoted_words: 11215
major_words: 11248
minor_collections: 54
major_collections: 3
forced_major_collections: 0
heap_words: 40960
top_heap_words: 40960
mean_space_overhead: 13567.659682
```

25k minor GCs in `spectralnorm_par` vs 54 in `fib_par2`. However, when I increased the size of the minor heap, while brought down the number of minor GCs, the running time remained the same.

In order to make progress on this manual page, I'm going to edit the paragraph as follows:

```
On the author's machine, the program scales reasonably well up to 4 domains but performs worse with 8 domains. Recall that the machine only has 4 physical cores. Debugging and fixing this performance issue is beyond the scope of this tutorial.
```

Reply:
LGTM